### ISSUE TITLE
Bean Machine (BMG) installation errors on M1/M2 Macs

### ISSUE THREAD

**Issue by jpchen** [facebookresearch/beanmachine#1400]
In Python 3.7 on Mac 12.3 running the M1 Pro chips, with C++17 and C++20 using
```
Apple clang version 13.1.6 (clang-1316.0.21.2)
Target: arm64-apple-darwin21.4.0`
```
BMG fails to compile with a variety of deprecation errors errors coming from eigen 3.3.* (snipped):
```
...
    /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/type_traits:4009:34: note: 'result_of<std::equal_to<unsigned long long> (const unsigned long long &, const unsigned long long &)>' has been explicitly marked deprecated here
    template <class _Callable> class _LIBCPP_DEPRECATED_IN_CXX17 result_of;
                                     ^
    /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__config:1066:39: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
    #  define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                          ^
    /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/v1/__config:1043:48: note: expanded from macro '_LIBCPP_DEPRECATED'
    #    define _LIBCPP_DEPRECATED __attribute__ ((deprecated))
                                                   ^
    4 errors generated.
```

[GCC is not supported on the new Macs](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=96168), so further investigation is warranted. The temporary fix is to remove the `-WError` flag in `setup.py`.

**Comment by horizon-blue** [comment]
I don't have a M1 mac at moment to test the built, though the deprecation warning seems to be a generic Eigen thing rather than specific to M1, which should be resolved by using Eigen 3.4 from `conda-forge` (#1394)

**Comment by jpchen** [comment]
Using C++20 and upgrading to 3.4 per your suggestion worked, thanks! (The other combinations of compilers with upgrading/downgrading eigen has a different set of deprecation errors for some reason).
